source/h2_frames.c

```diff
@@ -332,714 +256,829 @@ void aws_h2_frame_encoder_clean_up(struct aws_h2_frame_encoder *encoder) {
-    struct aws_h2_frame_encoder *encoder,
-    struct aws_byte_buf *output) {
-    AWS_PRECONDITION(frame);
-    AWS_PRECONDITION(encoder);
-    AWS_PRECONDITION(output);
-
-    const size_t output_init_len = output->len;
-
-    /* Calculate length & flags */
-    size_t length = 0;
-    if (aws_h2_frame_header_block_get_encoded_length(&frame->header_block, encoder, &length)) {
-        goto compression_error;
+    /* Max amount of body we can fit in the payload*/
+    size_t max_body;
+    if (aws_sub_size_checked(max_payload, payload_overhead, &max_body) || max_body == 0) {
+        goto handle_waiting_for_more_space;
     }
 
-    uint8_t flags = 0;
-    if (frame->end_stream) {
-        flags |= AWS_H2_FRAME_F_END_STREAM;
-    }
-    if (frame->end_headers) {
-        flags |= AWS_H2_FRAME_F_END_HEADERS;
-    }
-    if (frame->pad_length) {
-        flags |= AWS_H2_FRAME_F_PADDED;
-        length += 1 + frame->pad_length;
-    }
-    if (frame->has_priority) {
-        flags |= AWS_H2_FRAME_F_PRIORITY;
-        length += s_frame_priority_settings_size;
-    }
+    /* Use a sub-buffer to limit where body can go */
+    struct aws_byte_buf body_sub_buf =
+        aws_byte_buf_from_empty_array(output->buffer + output->len + bytes_preceding_body, max_body);
 
-    /* Write the header data */
-    if (s_frame_prefix_encode(&frame->base, length, flags, output)) {
-        goto write_error;
+    /* Read body into sub-buffer */
+    if (aws_input_stream_read(body_stream, &body_sub_buf)) {
+        goto error;
     }
 
-    /* Write pad length */
-    if (frame->pad_length) {
-        if (!aws_byte_buf_write_u8(output, frame->pad_length)) {
-            goto write_error;
-        }
+    /* Check if we've reached the end of the body */
+    struct aws_stream_status body_status;
+    if (aws_input_stream_get_status(body_stream, &body_status)) {
+        goto error;
     }
-    /* Write priority */
-    if (frame->has_priority) {
-        if (s_frame_priority_settings_encode(&frame->priority, output)) {
-            goto write_error;
+
+    if (body_status.is_end_of_stream) {
+        *body_complete = true;
+        if (body_ends_stream) {
+            flags |= AWS_H2_FRAME_F_END_STREAM;
         }
-    }
-    /* Write data */
-    if (aws_h2_frame_header_block_encode(&frame->header_block, encoder, output)) {
-        goto compression_error;
-    }
-    /* Write padding */
-    for (size_t i = 0; i < frame->pad_length; ++i) {
-        if (!aws_byte_buf_write_u8(output, 0)) {
-            goto write_error;
+    } else {
+        if (body_sub_buf.len == 0) {
+            /* This frame would have no useful information, don't even bother sending it */
+            goto handle_nothing_to_send_right_now;
         }
     }
 
-    return AWS_OP_SUCCESS;
-
-write_error:
-    output->len = output_init_len;
-    return aws_raise_error(AWS_H2_ERR_FRAME_SIZE_ERROR);
+    /*
+     * Write in the other parts of the frame.
+     */
+    bool all_wrote = true;
```

Comment:
I'm ashamed this even bothers me.... but

bool all_written

Reply:
I was so happy to read this comment, after all the trivial feedback I throw at everyone 🤣

I ended up naming it `writes_ok`
I tried `all_written` and realized it looked like I was tracking whether I was all done encoding (`all_wrote` was also misleading that way)